src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2693 @@
+
+ > message Wrapper {
+ >    bytes value = 1
+ > }
+
+ where the name @Wrapper@ can be arbitrary, but the @value@ field must exist,
+ and it must have type @bytes@.
+-}
+deMkSerialize ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        BS.ByteString
+    , S.Serialize a
+    ) =>
+    b ->
+    Maybe a
+deMkSerialize val = case S.decode $ val ^. ProtoFields.value of
+    Left _ -> Nothing
+    Right v -> v
+
+-- |Like 'deMkSerialize' above, but for Word64 fields.
+deMkWord64 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word64
+    ) =>
+    b ->
+    Word64
+deMkWord64 val = coerce $ val ^. ProtoFields.value
+
+-- |Like 'deMkWord64' above, but for Word32 fields instead.
+deMkWord32 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Word32
+deMkWord32 val = coerce $ val ^. ProtoFields.value
+
+-- |Like 'deMkWord64' above, but for Word16 fields instead.
+deMkWord16 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Maybe Word16
+deMkWord16 val =
+    if v <= 65_535
+        then return $ fromIntegral v
+        else Nothing
+  where
+    v = val ^. ProtoFields.value
+
+-- |Like 'mkWord32', but for Word8 fields instead.
+deMkWord8 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Maybe Word8
+deMkWord8 val =
+    if v <= 255
+        then return $ fromIntegral v
+        else Nothing
+  where
+    v = val ^. ProtoFields.value
+
+{- |A helper class analogous to something like Aeson's FromJSON.
+ It exists to make it more manageable to convert the Protobuf types to
+ their internal Haskell type equivalents.
+-}
+class FromProto a where
+    -- |The corresponding Haskell type.
+    type Output' a
+
+    -- |A conversion function from the protobuf type to its Haskell equivalent.
+    fromProto :: a -> Maybe (Output' a)
+
+fromProtoError :: Maybe a
+fromProtoError = Nothing
+
+-- |A block hash input
+data BlockHashInput = Best | LastFinal | Given !BlockHash
+
+-- |An account identifier input
+type AccountIdentifierInput = AccountIdentifier
+
+instance FromProto Proto.AccountIndex where
+    type Output' Proto.AccountIndex = AccountIndex
+    fromProto = return . AccountIndex . deMkWord64
+
+instance FromProto Proto.AccountAddress where
+    type Output' Proto.AccountAddress = AccountAddress
+    fromProto = deMkSerialize
+
+instance FromProto Proto.CredentialRegistrationId where
+    type Output' Proto.CredentialRegistrationId = (CredentialRegistrationID, RawCredentialRegistrationID)
+    fromProto crId = do
+        raw <- deMkSerialize crId
+        nonRaw <- credIdFromRaw raw
+        return (nonRaw, raw)
+      where
+        credIdFromRaw (RawCredentialRegistrationID fbs) =
+            case S.decode (FBS.toByteString fbs) of
+                Left _ -> Nothing
+                Right c -> c
```

Comment:
Do you want the client to fail with `error` here if the input is invalid? If that is the case, what's the point of wrapping the computation in a monad then?

Reply:
I am not sure what you are asking. I am just proposing you add a function `credIdFromRaw` to the same place where `unsafeCredIdfromRaw` is, instead of having multiple copies in the client project. That functioni should return Maybe CredId.

Reply:
I was just asking due to the ambiguity "this" and "here", but I completely agree that moving it to `Types.hs` is the right thing to do. Thanks.